packages/app/src/components/DBTraceWaterfallChart.tsx

```diff
@@ -22,233 +27,338 @@ type SpanRow = {
-    useOffsetPaginatedQuery(
-      useMemo(
-        () => ({
-          ...config,
-          dateRange: [focusDate, dateRange[1]],
-          dateRangeStartInclusive: false,
-          orderBy: [
-            {
-              valueExpression: traceTableModel.timestampValueExpression,
-              ordering: 'ASC',
-            },
-          ],
-        }),
-        [
-          config,
-          focusDate,
-          dateRange,
-          traceTableModel.timestampValueExpression,
-        ],
-      ),
-    );
-
-  const data = useMemo(() => {
-    return {
-      meta: beforeSpanData?.meta ?? afterSpanData?.meta,
-      data: [
-        ...(beforeSpanData?.data ?? []),
-        ...(afterSpanData?.data ?? []),
-      ].map(d => {
-        d.HyperDXEventType = 'span';
-        return d;
-      }) as SpanRow[],
-    }; // TODO: Type useOffsetPaginatedQuery instead
+    useFetchingData({
+      config,
+      dateRangeStartInclusive: false,
+      dateRange: [focusDate, dateRange[1]],
+    });
+  isFetching = isFetching || isBeforeSpanFetching || isAfterSpanFetching;
+  const meta = beforeSpanData?.meta ?? afterSpanData?.meta;
+  const rowWhere = useRowWhere({ meta, aliasMap: alias });
+  const rows = useMemo(() => {
+    // Sometimes meta has not loaded yet
+    // DO NOT REMOVE, useRowWhere will error if no meta
+    if (!meta || meta.length === 0) return [];
+    const concatData = [
+      ...(beforeSpanData?.data ?? []),
+      ...(afterSpanData?.data ?? []),
+    ].map(d => {
+      d.HyperDXEventType = 'span';
+      return d;
+    }) as SpanRow[];
+    return concatData.map((cd: SpanRow) => ({
+      ...cd,
+      id: rowWhere(omit(cd, ['HyperDXEventType'])),
+      type,
+    }));
   }, [afterSpanData, beforeSpanData]);
 
-  const isSearchResultsFetching = isBeforeSpanFetching || isAfterSpanFetching;
-
   return {
-    data,
-    isFetching: isSearchResultsFetching,
-    aliasMap,
+    rows,
+    isFetching,
   };
 }
 
 // TODO: Optimize with ts lookup tables
 export function DBTraceWaterfallChartContainer({
   traceTableModel,
+  logTableModel,
   traceId,
   dateRange,
   focusDate,
   onClick,
   highlightedRowWhere,
 }: {
   traceTableModel: TSource;
+  logTableModel?: TSource;
   traceId: string;
   dateRange: [Date, Date];
   focusDate: Date;
-  onClick?: (rowWhere: string) => void;
+  onClick?: (rowWhere: { id: string; type: string }) => void;
   highlightedRowWhere?: string | null;
 }) {
-  const { data, isFetching, aliasMap } = useSpansAroundFocus({
-    traceTableModel,
+  const { rows: traceRowsData, isFetching: traceIsFetching } =
+    useSpansAroundFocus({
+      tableModel: traceTableModel,
+      focusDate,
+      dateRange,
+      traceId,
+    });
+  const { rows: logRowsData, isFetching: logIsFetching } = useSpansAroundFocus({
+    // search data if logTableModel exist
+    // search invliad date range if no logTableModel(react hook need execute no matter what)
```

Comment:
```suggestion
    // search invalid date range if no logTableModel(react hook need execute no matter what)
```

Also: you can pass in enable flags into queries, typically we expose them upwards (looks like you can pass `enable` into the second arg of `useOffsetPaginatedQuery`). If this doesn't fire a query, maybe that's something we can skip for now but if we fire a silly query we should just disable it.

Reply:
Isn't caching results if the query is the same a good thing? Not sure I'm following.

Reply:
Because in this component, we have two sources, `traceSource` and `logSource`, but when user does not provide `logSource`, we use  `traceSource` as fake `logSource` since we still need to build query and call react use function.

And caching makes we get duplicate data even with `enabled: false`, this solution just make sure we use different query so `enable` flag can be used when calling use function.